cli/lsp/config.rs

```diff
@@ -139,25 +146,189 @@ impl WorkspaceSettings {
+                        }
+                      }
+                    }
+                    _ => {
+                      match serde_json::from_value::<SpecifierSettings>(value) {
+                        Ok(specifier_settings) => {
+                          let (specifier, uri) =
+                            specifier_uri_map[i - 1].clone();
+                          settings
+                            .specifiers
+                            .insert(specifier, (uri, specifier_settings));
+                        }
+                        Err(err) => {
+                          error!(
+                            "Error converting specifier settings: {}",
+                            err
+                          );
+                        }
+                      }
+                    }
+                  }
+                }
+              }
+            }
+            Some(ConfigRequest::Specifier(specifier, uri)) => {
+              if let Ok(value) = client
+                .configuration(vec![lsp::ConfigurationItem {
+                  scope_uri: Some(uri.clone()),
+                  section: Some(SETTINGS_SECTION.to_string()),
+                }])
+                .await
+              {
+                match serde_json::from_value::<SpecifierSettings>(
+                  value[0].clone(),
+                ) {
+                  Ok(specifier_settings) => {
+                    settings_ref
+                      .lock()
+                      .unwrap()
+                      .specifiers
+                      .insert(specifier, (uri, specifier_settings));
+                  }
+                  Err(err) => {
+                    error!("Error converting specifier settings: {}", err);
+                  }
+                }
+              } else {
+                error!(
+                  "Error retrieving settings for specifier: {}",
+                  specifier
+                );
+              }
+            }
+          }
+        }
+      })
+    });
+
+    Self {
+      client_capabilities: ClientCapabilities::default(),
+      root_uri: None,
+      settings,
+      tx,
+    }
+  }
+
+  pub fn get_workspace_settings(&self) -> WorkspaceSettings {
+    self.settings.lock().unwrap().workspace.clone()
+  }
+
+  /// Set the workspace settings directly, which occurs during initialization
+  /// and when the client does not support workspace configuration requests
+  pub fn set_workspace_settings(&self, value: Value) -> Result<(), AnyError> {
+    let workspace_settings = serde_json::from_value(value)?;
+    self.settings.lock().unwrap().workspace = workspace_settings;
+    Ok(())
+  }
+
+  pub fn snapshot(&self) -> ConfigSnapshot {
+    ConfigSnapshot {
+      client_capabilities: self.client_capabilities.clone(),
+      root_uri: self.root_uri.clone(),
+      settings: self.settings.lock().unwrap().clone(),
+    }
   }
 
   pub fn specifier_enabled(&self, specifier: &ModuleSpecifier) -> bool {
```

Comment:
Can I suggest changing these to associated methods that take a MutexGuard? I.e., requiring prove that you hold the lock when you call them?

Imagine a future change where you acquire the lock but forget `specifier_enabled()` does too and then you call it: deadlock.

Reply:
I get the concept, but am not sure what that would look like, any examples you can point me to?

Reply:
Something like this:
```rust
pub fn specifier_enabled(g: &MutexGuard<Self>, specifier: &ModuleSpecifier) -> bool {
  // ...
}
```
And you'd call like this:
```rust
let g = config.lock().unwrap();
// ...
if Config::specifier_enabled(&g, specifier) {
  // ...
}
// ...

Reply:
Ok! I thought there was some magic, like I know you are have an `Arc<Self>`.

Hrmphh... That makes things a lot more complicated, like having to make `settings` public and having to acquire the lock.  What about if I `try_lock` and propagate the result that can be handled by the caller.  That would be cleaner.  The main goal is to just not have a blocking acquire.

Reply:
Hrmppph... I don't like that either... as it sort of makes a mess of things and can run into issues... I know we have avoided RwLocks before, but there is really only one part of the code now that can possibly write to settings (actually 2, but they both a very controlled) and having multiple reads is perfectly acceptable in this use case.  So I think a RwLock would actually be much better suited (and safer) here and for something that could be a high-frequency, like checking if enable or "snapshotting" it feels an appropriate use.

Reply:
> That makes things a lot more complicated, like having to make settings public and having to acquire the lock.

That's the idea :-) - it moves the decision of when and where to lock to the caller. The caller knows that better than the callee.

There's currently an implicit contract between caller and callee that says "don't lock before you call me because I'm going to." With a MutexGuard, you make that contract explicit.